### PR TITLE
Docs infra: ambiguous letter-casing doc filenames

### PR DESCRIPTION
**Pull request by petebacondarwin**
The motivation for this PR is actually to support upgrading AIO to Angular CLI v12 (and so Webpack 5).

### CODE REVIEW

**Comment by petebacondarwin** [comment]
Although this PR will result in some URL changes, the original URLs should land on the disambiguation page leading the reader to the correct place anyway.

File: aio/tools/transforms/templates/disambiguator.template.html [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+<div class="content">
+  There are multiple pages that match this path:
+  <ul>
+    {% for ambiguousDoc in doc.docs %}<li>
+      <a href="{$ ambiguousDoc.path $}">{$ ambiguousDoc.id $}</a>
+      {$ ambiguousDoc.shortDescription | marked $}
+    </li>{% endfor %}
+  </ul>
+</div>
+{% endblock %}
```
Comment (atscott) [inline aio/tools/transforms/templates/disambiguator.template.html:15]: I'm not familiar with the docs infrastructure. Can you link to an example where this is shown?

**Review by atscott — APPROVED** [review]

File: aio/tools/transforms/templates/disambiguator.template.html (hunk 1 above)
Comment (petebacondarwin) [reply aio/tools/transforms/templates/disambiguator.template.html:15]: Here you go: https://pr41788-39ca413.ngbuilds.io/api/core/injector

Comment (petebacondarwin) [reply aio/tools/transforms/templates/disambiguator.template.html:15]: Screenshot 2021-04-23 at 17 12 25

File: aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * are disambiguated.
+ */
+module.exports = function disambiguateDocPathsProcessor(log) {
+  return {
+    $runAfter: ['paths-computed'],
+    $runBefore: ['rendering-docs'],
+    $process(docs) {
+      // Collect all the ambiguous docs, whose outputPath is are only different by casing.
+      const ambiguousDocMap = new Map();
+      for(const doc of docs) {
```
Comment (gkalpak) [inline aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js:15]:
```suggestion
      for (const doc of docs) {
```

File: aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        const outputPath = doc.outputPath.toLowerCase();
+        if (!ambiguousDocMap.has(outputPath)) {
+          ambiguousDocMap.set(outputPath, []);
+        }
+        const ambiguousDocs = ambiguousDocMap.get(outputPath);
+        ambiguousDocs.push(doc);
+      }
+
+      // Create a disambiguator doc for each set of such ambiguous docs,
+      // and update the ambigous docs to have unique `path` and `outputPath` properties.
```
Comment (gkalpak) [inline aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js:28]:
```suggestion
      // and update the ambiguous docs to have unique `path` and `outputPath` properties.
```

File: aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      for(const outputPath of ambiguousDocMap.keys()) {
```
Comment (gkalpak) [inline aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js:29]:
```suggestion
      for (const outputPath of ambiguousDocMap.keys()) {
```

You could also use the entries directly (and not have to retrieve the value):

```suggestion
      for (const [outputPath, ambiguousDocs] of ambiguousDocMap) {
```

File: aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        const doc = ambiguousDocs[0];
+        const path = doc.path;
+        const id = `${doc.id.toLowerCase()}-disambiguator`;
+        const title = `${doc.id.toLowerCase()} (disambiguation)`;
+        const aliases = [id];
+        docs.push({ docType: 'disambiguator', id, title, aliases, path, outputPath, docs: ambiguousDocs });
+
+        // Update the paths
+        let count = 0;
+        for(const doc of ambiguousDocs) {
```
Comment (gkalpak) [inline aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js:46]:
```suggestion
        for (const doc of ambiguousDocs) {
```

File: aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+          count += 1;
+        }
+      }
+    }
+  };
+};
+
+function convertPath(outputPath, count) {
+  // Add the counter before any extension
+  return outputPath.replace(/(\.[^.]*)?$/, `-${count}$1`).toLowerCase();
```
Comment (gkalpak) [inline aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js:58]: Thought: Would it be better to not use `toLowerCase()` (and thus keep the casing (for browsers and case-sensitive OSes)?

```suggestion
  return outputPath.replace(/(\.[^.]*)?$/, `-${count}$1`);
```

File: aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.spec.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
+const testPackage = require('../../helpers/test-package');
+const Dgeni = require('dgeni');
+
+describe('disambiguateDocPaths processor', () => {
+  let dgeni, injector, processor, docs;
+
+  beforeEach(function() {
```
Comment (gkalpak) [inline aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.spec.js:7]: You forgot the `=>` and there seems to be some stray random string (`function`) before the `()`.
(Strange that the JS parser didn't complain 😛)

File: aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.spec.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+  it('should be part of the dgeni package', () => {
+    expect(processor).toBeDefined();
+  });
+
+  it('should create `disambiguator` documents for docs that have ambiguous outputPaths', () => {
+    const numDocs = docs.length;
+    processor.$process(docs);
+    expect(docs.length).toEqual(numDocs + 2);
+    expect(docs[docs.length-2]).toEqual({
```
Comment (gkalpak) [inline aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.spec.js:29]:
```suggestion
    expect(docs[docs.length - 2]).toEqual({
```

😇

File: aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.spec.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+      docType: 'disambiguator',
+      id: 'test-doc-disambiguator',
+      title: 'test-doc (disambiguation)',
+      aliases: ['test-doc-disambiguator'],
+      path: 'test/doc',
+      outputPath: 'test/doc.json',
+      docs: [docs[0], docs[1]],
+    });
+    expect(docs[docs.length-1]).toEqual({
```
Comment (gkalpak) [inline aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.spec.js:38]:
```suggestion
    expect(docs[docs.length - 1]).toEqual({
```

😇

**Review by gkalpak — CHANGES_REQUESTED** [review]
Awesome work! (You make look too easy, which is demotivating :stuck_out_tongue:)

One comment:
We could improve how the disambiguation pages show up in search results: 
disambiguation-search

Ideally we should not show them at all (but that might be a bigger change). Maybe we could at least give them a [label + symbol][1] :grin:

Commit message typos:
  - `initlalized` --> `initialized` (in the first commit)
  - `ambigous` --> `ambiguous` (in the second and third commits)

[1]: https://github.com/angular/angular/blob/73105aa7d6cf1116a6a9f103bd074f65db7584fe/aio/src/styles/_constants.scss#L69-L127

File: aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js (hunk 2 above)
Comment (petebacondarwin) [reply aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js:15]: We really need to get doc-formatting up and running for the doc-gen code.

File: aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js (hunk 3 above)
Comment (petebacondarwin) [reply aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js:28]: and spell checking :-)

File: aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js (hunk 6 above)
Comment (petebacondarwin) [reply aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js:58]: Yeah - we can try that.

**Comment by petebacondarwin** [comment]
The disambiguator pages are gone from the search.

https://pr41788-887cbb3.ngbuilds.io/?search=inject

**Review by gkalpak — APPROVED** [review]
:100: 🚀 🏆 

(BTW, there is still an `ambigous` occurrence in the third (non-fixup) commit.)

File: aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js [hunk 10]
```diff
@@ -0,0 +1,58 @@
+/**
+ * @dgProcessor disambiguateDocPathsProcessor
+ * @description
+ *
+ * Ensures that docs that have the same path, other than case changes,
+ * are disambiguated.
```
Comment (jelbourn) [inline aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js:6]: Could you expand this comment with an example of a case that would require disambiguation?

**Review by jelbourn — APPROVED** [review]
LGTM

Reviewed-for: docs-packaging-and-releasing

File: aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js (hunk 10 above)
Comment (petebacondarwin) [reply aio/tools/transforms/angular-base-package/processors/disambiguateDocPaths.js:6]: Will do! Thanks for the review @jelbourn

**Review by atscott — APPROVED** [review]
reviewed-for: public-api

**Review by jelbourn — APPROVED** [review]
LGTM

Reviewed-for: public-api

**Review by jessicajaniuk — APPROVED** [review]
LGTM :cookie:

reviewed-for: public-api

**Pull request closed without merging**